### PR TITLE
feat: rework api to use enum flags, add filter procs

### LINKED ISSUE #8: feature request enable huge speedups: allow passing a filter on intermediate directories
feature request enable huge speedups: allow passing a filter on intermediate directories; the code will avoid recursing on directories that fail fillterDir

```nim
proc myFillterDir(dir: string): bool =
  return dir.count("/") < 3 # example 1: only go to a certain depth
  return !dir.match(re"bad_pattern") # example 2: avoid recursing on some directories

for path in walkGlob("./**/*.nim", fillterDir = myFillterDir):
  echo path
```

### LINKED ISSUE #11: [RFC] API simplification
The current API is both too limited (no way to only output symlinks, or only dirs, etc) and fixing these using existing approach (via new flag onlySymlinks, excludeSymlinks etc) is just not a good design.

I propose the following simplifications, based on principle of separation of concerns: deprecate some of the existing flags in favor of using a filter (cf my suggestion in https://github.com/citycide/glob/pull/10#issuecomment-403165060 to distinguish `filterDescend` and `filterYield`).

* deprecate `includeHidden`
=> the user can now control whether he means to filter out such paths via filterDescend or filterYield, eg:
```nim
walkGlobKinds(".").filter(x=>not x.path.isHidden) # allow yielding ./bar/.git/foo but not ./bar/.gitconfig
walkGlobKinds(".", filterDescend= dir=>not dir.isHidden) # disallow ./bar/.git/foo and ./bar/.gitconfig
```
Also, current `isHidden` doesn't allow one to only yield hidden paths, so it's not very flexible.

* deprecate `includeDirs`
right now, there's no way to only produce dirs (cf unix cmd `find . -type d`), only exclude dirs, so `includeDirs` is not very flexible.
Instead this would work:

```nim
walkGlobKinds(".").filter(x=>x.kind.isDir)
```

* (once `toSeq` becomes a proc, to allow UFCS, I have a work in progress on that) deprecate `listGlob` in favor of:
```nim
let a = listGlob(args)
=>
let a = walkGlobKinds(args).toSeq
```

### LINKED ISSUE #12: controlling case (in-)sensitivity in glob
in python, ignoring case seems a bit awkard, eg: https://stackoverflow.com/questions/8151300/ignore-case-in-glob-on-linux
what would be recommended way to have case insensitive glob with this library?
if too awkward to use, maybe we should have a flag to control that?

### LINKED ISSUE #13: walkGlobKinds(".") excludes "." ; needs to be documented: consistent with `python glob` but inconsistent with unix `find .`

### LINKED ISSUE #17: symlinks should not be followed by default; can cause infinite recursion or list too many things because of just 1 bad symlink
```
tree
.
└── foo
    └── bar
        └── foo_link -> ../
```

NOTE: these occur in practice, eg on OSX in app folders IIRC.

```nim
import glob,strutils,sequtils
echo toSeq(glob.walkGlob("foo", includeDirs=true)).join("\n")
```

```
nim c -o:app test.nim
```

```
foo/bar
foo/bar/foo_link
foo/bar/foo_link/bar
foo/bar/foo_link/bar/foo_link                                                                                                                                                                                                                      foo/bar/foo_link/bar/foo_link/bar
foo/bar/foo_link/bar/foo_link/bar/foo_link
foo/bar/foo_link/bar/foo_link/bar/foo_link/bar
foo/bar/foo_link/bar/foo_link/bar/foo_link/bar/foo_link
foo/bar/foo_link/bar/foo_link/bar/foo_link/bar/foo_link/bar
foo/bar/foo_link/bar/foo_link/bar/foo_link/bar/foo_link/bar/foo_link
etc... (82 entries)
```

Note: curiously, this terminates gracefully without error after 82 lines (maybe an OS limit on path depth, since I'm seeing same behavior with D and python):
```
./app && echo ok || echo errror
ok
./app | wc -l
82
```

- [ ] there should be an option `followSymlinks = false` and indeed symlinks shouldn't be followed by default

- [ ] Note, https://dlang.org/phobos/std_file.html#dirEntries uses `followSymlink = true` but I think that's the wrong default ; likewise python glob follows and i'm not even seeing a way to not follow symlinks...); 

Unix find has the correct default (not follow symlinks), and only follow symlinks if -L is provided, see https://unix.stackexchange.com/a/31121/211423 for more on that rationale.
(either symlink recursion as shown above or a single symlink deep down pointing to ~ or / will cause a ton of entries.)

- [ ] glob should also detect and avoid infinite recursion. Note that D's dirEntries and python's glob currently don't seem to detect this case, but unix find does handle that correctly.


NOTE: alternative would be to not have an option `followSymlinks` and  follow symlinks, and let user filter out symlinks via `filterDescend ` from https://github.com/citycide/glob/issues/8 ; however that's a common enough use case that it should have its own option IMO; also it's indeed a tricky thing to handle (eg infinite recursion) so right here in glob is the right place to handle that.

### PR DESCRIPTION
**Pull request by haltcase**
Closes #8 ([ref](https://github.com/citycide/glob/blob/3e72cde6d2ebaae6cd4d8d50e9f9bd61c70f12f8/src/glob.nim#L351)), closes #10 ([ref](https://github.com/citycide/glob/blob/3e72cde6d2ebaae6cd4d8d50e9f9bd61c70f12f8/src/glob.nim#L344)), closes #11, closes #13 ([ref](https://github.com/citycide/glob/blob/3e72cde6d2ebaae6cd4d8d50e9f9bd61c70f12f8/src/glob.nim#L295-L296)), closes #17 ([ref](https://github.com/citycide/glob/blob/3e72cde6d2ebaae6cd4d8d50e9f9bd61c70f12f8/src/glob.nim#L328-L338)), closes #12 ([ref](https://github.com/citycide/glob/blob/b5dc6c68a4757e86f369f39b1020b4056f5c5a44/src/glob.nim#L185))

### API changes

Options (like `includeDirs` or `includeHidden`) are no longer provided as individual boolean parameters and are instead part of a `set[GlobOption]`, the default for which is exported as `defaultGlobOptions`. You can alter this like you would any other set using unions and intersections:

```nim
import glob
import sequtils

echo toSeq(walkGlob("src", options = defaultGlobOptions + {Hidden, FollowLinks} - {Files}))

# or replace the default options completely
echo toSeq(walkGlob("src", options = {Hidden}))
# @[... only hidden files under src/ ...]
```

#### `listGlob` removal

The `seq`-returning proc `listGlob` is also removed to encourage the use of the iterators `walkGlob` or `walkGlobKinds`. It's also easy enough to get a `seq` from these using `sequtils.toSeq` from Nim's stdlib that it was mostly unnecessary, and this reduces the number of signatures we need to keep in sync.

#### Option flags

The new `GlobOption` enum is defined as:

```nim
type
  GlobOption* {.pure.} = enum
    ## Flags that control the behavior or results of the file system iterators. See
    ## `defaultGlobOptions <#defaultGlobOptions>`_ for some usage & examples.
    Absolute, IgnoreCase, NoExpandDirs, FollowLinks,  ## iterator behavior
    Hidden, Files, Directories, FileLinks, DirLinks   ## to yield or not to yield
```

Several of these are equivalent to parameters in the current API, but a few of them are new:

| current                |  flag                      |  meaning                                                 |
| ---------------------- | -------------------------- | -------------------------------------------------------- |
| `relative = false`     | `GlobOption.Absolue`       | yield paths as absolute rather than relative to root     |
| `expandDirs = false`   | `GlobOption.NoExpandDirs`  | if pattern is a directory don't treat it as `<dir>/**/*` |
| `includeHidden = true` | `GlobOption.Hidden`        | yield hidden files or directories                        |
| `includeDirs = true`   | `GlobOption.Directories`   | yield directories                                        |
| N/A                    | `GlobOption.IgnoreCase`    | matching will ignore case differences                    |
| N/A                    | `GlobOption.Files`         | yield files                                              |
| N/A                    | `GlobOption.DirLinks`      | yield links to directories                               |
| N/A                    | `GlobOption.FileLinks`     | yield links to files                                     |
| N/A                    | `GlobOption.FollowLinks`   | recurse into directories through links                   |

So this new options set allows for more granular things like only returning links:

```nim
for path in walkGlob(".", options = {FileLinks, DirLinks}):
  # only links!
  discard
```

#### More advanced filtering

When the options flags aren't enough, you can do a lot more with the optional `filterYield` or `filterDescend` procs (subsuming #10) to decide what gets yielded and what directories are traversed.

```nim
type
  FilterDescend* = (path: string) -> bool
  FilterYield* = (path: string, kind: PathComponent) -> bool
```

Both of these receive a path as their first parameter (relative or absolute, depending on if `Absolute in options`).

`filterDescend` is called whenever the iterator comes across a directory that is about to be recursed into. If you return false, the iterator will abandon that recursion. Any path received by a `filterDescend` proc is guaranteed to be a directory or a link to a directory.

`filterYield` is called whenever an item (file, directory, or link) is about to be yielded. If you return false, the iterator will _not_ yield the item.

These procs are called _after_ options flags are checked, so if `Files notin options` then no paths pointing to files will be passed to `filterYield`.

### CODE REVIEW

**Comment by timotheecour** [comment]
Cool!
@citycide you should use ospaths.FileSystemCaseSensitive (it's buggy because of https://github.com/nim-lang/Nim/issues/8349 but I can send a PR to Nim soon) instead of re-defining the logic in `IgnoreCase` (note: you did the same bug as in https://github.com/nim-lang/Nim/issues/8349, lol!)

**Comment by citycide** [comment]
> you should use ospaths.FileSystemCaseSensitive 

I did use that for part of it, maybe it can be used more but I'd wait for that bug to get fixed.

> what's the rationale for ``2f8d1a2 feat: flip `ExpandDirs` option `toNoExpandDirs``` ?

I noticed when putting together examples that it was easier to have fewer defaults because it makes it fairly simple when completely replacing them. I know it's pretty easy to just modify the defaults but this seems like a good balance. Generally I'd agree with you and wouldn't use a negated option but the ergonomics on this felt better.

File: tests.nim [hunk 1]
```diff
@@ -337,6 +338,15 @@ suite "pattern walking / listing":
         "temp" / "shallow.nim"
       ])
 
+    test "`IgnoreCase` enables case insensitive matching":
+      let o = defaultGlobOptions + {IgnoreCase}
+      check seqsEqual(toSeq(walkGlob("TEMP/**", options = o)), @[
+        "temp" / "deep" / "dir" / "file.nim",
```
Comment (timotheecour) [inline tests.nim]: nit: maybe more readable via:
`"temp/deep/dir/file.nim".unixToNativePath` ?

**Comment by citycide** [comment]
> nit: IgnoreCase => CaseInsensitive ? to match existing terminology in makeCaseInsensitive

That's an internal proc, shouldn't need to worry much about consistency between it and a public flag.

File: tests.nim (hunk 1 above)
Comment (citycide) [reply tests.nim]: Hmm might even be nice to put a template in here to do `p"temp/deep/dir/file.nim"`

**Comment by citycide** [comment]
> wouldn't it make more sense to add a flag CaseSensitive (defaulting to true) in https://github.com/citycide/glob/blob/master/src/glob/regexer.nim instead of transforming pattern in makeCaseInsensitive ?

That's what I did for the parser:

https://github.com/citycide/glob/blob/b5dc6c68a4757e86f369f39b1020b4056f5c5a44/src/glob/regexer.nim#L74

`makeCaseInsensitive` is there because there's no way on case sensitive systems (that I know of) to list files case insensitively. So we turn each character in the string into a character class if its lower and upper case forms, then pass that as a pattern to Nim's `os.walkPattern` to gather up all the matches.

> does FollowLinks take care of avoiding infinite recursion? wasn't clear from top-level msg

Yep, see here:

https://github.com/citycide/glob/blob/b5dc6c68a4757e86f369f39b1020b4056f5c5a44/src/glob.nim#L448-L456

> is UTF8 supported ? (eg in case of case sensitivity + other); whatever answer is, would be nice to specify in docs

Not currently, we can leave it as a follow on and add it to the docs for now. It might work outside case sensitivity though.

**Comment by timotheecour** [comment]
@citycide please see https://github.com/citycide/glob/pull/20 ; seems much simpler than https://github.com/citycide/glob/commit/7115d8d39d420beaf4bf4c3f167105183576c305 for handling case insensitivity

**Pull request merged**